### ISSUE TITLE
Hyper-V socket crash with 4.11 kernels

### ISSUE THREAD

**Issue by rn** [linuxkit/linuxkit#1848]
PR https://github.com/linuxkit/linuxkit/pull/1844 updated the experimental kernel patches for the 4.11.1 kernel to the latest from https://github.com/dcui/linux/commits/decui/hv_sock/v4.11/20170511 which fixed an initial issue. However running longer stress tests with more concurrency cause a kernel crash on Windows 10 Pro RS1 build 14393.1066 and a hang of the Linux VM on Windows 10 Pro Insider fast ring build 16193.1001. It's unclear if the Windows build is significant as the machines are different specs.

To reproduce:
- Building a LinuxKit image with the stress test tools from https://github.com/linuxkit/virtsock by typing `make linuxkit`.
- Copy `hvtest-efi.iso` and `./build/virtsock_stress.exe` to your Windows machine
- Copy [`LinuxKIt.ps1`](https://github.com/linuxkit/linuxkit/blob/master/scripts/LinuxKit.ps1) to your Windows machine
On the windows machine:
- Make sure the Hyper-V socket UUIDs are registered using [this script](https://github.com/linuxkit/virtsock/blob/master/scripts/setup.ps1). This is a one time operation.
- Create a Hyper-V VM and start it from an elevated powershell:
  `.\LinuxKit.ps1 -Create -IsoFile .\hvtest-efi.iso -Memory 2048 -CPUs 2; .\LinuxKit.ps1 -start`
- Connect to the serial console: `putty -serial  \\.\pipe\LinuxKitVM-com1`
- In the LinuxKitVM start the server process: `virtsock_stress -v 1 -s`
- In Powershell, get the ID for the VM: `(Get-VM LinuxKitVM).Id`
- In a powershell on the host start the client:
  `virtsock_stress.exe -v 1 -c <VM ID> -i 1000000 -p 10`

The last command starts the client which will use 10 threads to connect to the server with a connection per thread. Up to 1000000 connections will be made an random amount of data is transferred over each connection and echoed back by the server. The client is also comparing checksums of the data sent and received.

On the Insider build 16193.1001 this hangs after some 10000 connections have been made and the client eventually gets a receive timeout. The server in the Linux VM is locked (can't `ctrl-c`) until we exit the client.

On the RS1 build (14393.1066) there is a kernel crash after about 13000 connections:
```
[  317.821101] ------------[ cut here ]------------
[  317.895414] WARNING: CPU: 1 PID: 543 at lib/list_debug.c:56 __list_del_entry_                                                     valid+0x51/0x84
[  317.922774] list_del corruption. next->prev should be ffff9d87b0197078, but w                                                     as ffff9d87b6ce3178
[  317.945516] Modules linked in:
[  317.951779] CPU: 1 PID: 543 Comm: virtsock_stress Not tainted 4.11.1-linuxkit                                                      #1
[  317.961773] Hardware name: Microsoft Corporation Virtual Machine/Virtual Mach                                                     ine, BIOS Hyper-V UEFI Release v1.0 11/26/2012
[  317.961773] Call Trace:
[  317.961773]  ? dump_stack+0x5a/0x6f
[  317.961773]  ? __warn+0xd2/0xeb
[  317.961773]  ? warn_slowpath_fmt+0x5a/0x76
[  317.961773]  ? __list_del_entry_valid+0x51/0x84
[  317.961773]  ? __list_del_entry+0x9/0x1d
[  317.961773]  ? vsock_accept+0x16b/0x27e
[  317.961773]  ? add_wait_queue+0x3a/0x3a
[  317.961773]  ? SYSC_accept4+0x10b/0x1b9
[  317.961773]  ? SyS_futex+0xd1/0x158
[  317.961773]  ? entry_SYSCALL_64_fastpath+0x1a/0xa9
[  318.117012] ---[ end trace 66affefa1d1dbd3d ]---
[  318.132328] BUG: unable to handle kernel paging request at ffffb1b187be1004
[  318.148371] IP: hvs_stream_has_data+0x22/0x4f
[  318.158456] PGD 7f0ae067
[  318.158456] PUD 7f0af067
[  318.169820] PMD 77507067
[  318.169820] PTE 0
[  318.179915]
[  318.189999] Oops: 0000 [#1] SMP
[  318.200275] Modules linked in:
[  318.210337] CPU: 1 PID: 537 Comm: virtsock_stress Tainted: G        W       4                                                     .11.1-linuxkit #1
[  318.220461] Hardware name: Microsoft Corporation Virtual Machine/Virtual Mach                                                     ine, BIOS Hyper-V UEFI Release v1.0 11/26/2012
[  318.250740] task: ffff9d87b75ba280 task.stack: ffffb1b182e6c000
[  318.270897] RIP: 0010:hvs_stream_has_data+0x22/0x4f
[  318.281027] RSP: 0018:ffffb1b182e6fd30 EFLAGS: 00010246
[  318.291121] RAX: 0000000000000001 RBX: ffff9d87b0196d40 RCX: ffff9d87b75ba280
[  318.311320] RDX: ffffb1b187be1000 RSI: 0000000000000246 RDI: ffff9d87b0196d40
[  318.353684] RBP: 0000000000000001 R08: ffff9d87b0230400 R09: ffff9d87bd550280
[  318.373919] R10: 0000000000000000 R11: ffff9d87b75ba280 R12: 7fffffffffffffff
[  318.404151] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000004
[  318.424381] FS:  00007f3b58ac5eb0(0000) GS:ffff9d87b7a40000(0000) knlGS:00000                                                     00000000000
[  318.444527] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  318.460023] CR2: ffffb1b187be1004 CR3: 0000000070061000 CR4: 00000000003406a0
[  318.474838] Call Trace:
[  318.485127]  ? vsock_stream_recvmsg+0x16f/0x34a
[  318.495222]  ? core_sys_select+0x179/0x220
[  318.505296]  ? core_sys_select+0x1ce/0x220
[  318.515370]  ? add_wait_queue+0x3a/0x3a
[  318.535562]  ? sock_read_iter+0x80/0xab
[  318.545663]  ? __vfs_read+0x8c/0xc4
[  318.555728]  ? vfs_read+0x94/0xb0
[  318.565799]  ? SyS_read+0x5c/0x9d
[  318.565799]  ? entry_SYSCALL_64_fastpath+0x1a/0xa9
[  318.585973] Code: c3 0f 1f 44 00 00 31 c0 c3 0f 1f 44 00 00 48 8b 97 20 04 00                                                      00 b8 01 00 00 00 83 7a 38 00 75 37 4c 8b 42 28 49 8b 90 20 01 00 00 <8b> 72 04                                                      8b 0a 41 8b 90 30 01 00 00 41 89 c9 01 ca 41 29 f1 29
[  318.636432] RIP: hvs_stream_has_data+0x22/0x4f RSP: ffffb1b182e6fd30
[  318.656575] CR2: ffffb1b187be1004
[  318.656575] ---[ end trace 66affefa1d1dbd3e ]---
[  318.717071] Kernel panic - not syncing: Fatal exception
[  318.727136] Kernel Offset: 0x19000000 from 0xffffffff81000000 (relocation ran                                                     ge: 0xffffffff80000000-0xffffffffbfffffff)
[  318.727136] ---[ end Kernel panic - not syncing: Fatal exception
```

I've seen both the kernel crash as well as the hang on two successive runs (with restarting the VM in between, of course). So it seems easy to reproduce

**Comment by dcui** [comment]
Hi Rolf, do you still have the complete log? Here I see the crash happened in hvs_stream_has_data(). It would be good to see the messages before the crash.

In my testes, I only saw crashes in vsock_dequeue_accept()  and I think I have found a locking issue on dequeue/enqueue, and will need more time to make a fix.

**Comment by dcui** [comment]
Updates:
1. Made a proof-of-concept patch, which seems to be able to address the vsock_dequeue/enqueue issue (I need to figure out how to fix it in a correct way, probably using list_add_tail_rcu/list_del_rcu?) and I can't see the "Warning...lib/list_debug.c" any more. The patch is here:
https://github.com/dcui/linux/commit/a70eb589b14fc845e62dc0bb74444ab72b2ffc8e

2. Found vmbus_open() could get hang. Need further debugging.

3. In a 2-CPU VM, with hvstress -c <VM_ID> -p 10 (i.e. 10 parallel connect request), "top" shows the many udevd daemons are pretty busy. Need to address this issue.

4. It looks the recent patch (Drivers: hv: vmbus: Fix rescind handling) breaks hv_sock. Need further debugging:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20170519&id=54a66265d6754b37fb4baf1d970ec88a6225a988

**Comment by rneugeba** [comment]
Apologies for the delay. I have tried the new patch yet, but ran the test again with a debug kernel which has lockdep and a few other debug settings turned on: https://github.com/linuxkit/linuxkit/blob/master/kernel/kernel_config.debug

The first trace shows that there is list corruption detected during an `accept()` call. If you need the kernel symbols for the kernel used, you can do: 
```
docker pull linuxkit/kernel:4.11.2_dbg-71ff8b34d0a3f59072fe63fe233bca1f9ed3a3fd 
docker save -o /tmp/kernel.tar linuxkit/kernel:4.11.2_dbg-71ff8b34d0a3f59072fe63fe233bca1f9ed3a3fd
tar xvf /tmp/kernel.tar
tar xvf d2516c264ae0176d677ee155637c0e0547b5314251c68f8abd3caf36043e0400/layer.tar
```
and you should get the `vmlinux` binary


```
2017/05/25 09:58:24 [01885] ECHOED:      28178 bytes in    62.9657 ms
2017/05/25 09:58:24 [01879] ECHOED:      61354 bytes in   329.7187 ms
2017/05/25 09:58:24 [01888] ECHOED:      34903 bytes in     1.9221 ms
[  189.725479] ------------[ cut here ]------------
[  189.773384] WARNING: CPU: 1 PID: 655 at lib/list_debug.c:56 __list_del_entry_valid+0x55/0x8a
2017/05/25 09:58[  189.846339] list_del corruption. next->prev should be ffff8862680b5638, but was ffff8862665e6538
:24 [01889] ECHO[  189.943200] Modules linked in:
ED:      45632 b[  189.975533] CPU: 1 PID: 655 Comm: virtsock_stress Not tainted 4.11.2-linuxkit #1
ytes in    54.46[  189.985495] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v2.0 08/26/2016
45 ms[  189.985495] Call Trace:
[  190.128233]  dump_stack+0x82/0xb8
[  190.128233]  __warn+0xc2/0xdd

[  190.128233]  warn_slowpath_fmt+0x5a/0x76
[  190.182461]  ? _raw_spin_unlock_irqrestore+0x4e/0x5a
[  190.182461]  __list_del_entry_valid+0x55/0x8a
[  190.182461]  __list_del_entry+0xd/0x22
[  190.182461]  vsock_accept+0x179/0x299
2017/05/25 09:58[  190.182461]  ? add_wait_queue+0x44/0x44
:25 [01890] ECHO[  190.182461]  SYSC_accept4+0x127/0x1ea
[  190.182461]  ? entry_SYSCALL_64_fastpath+0x5/0xc2
ED:      34601 b[  190.182461]  ? trace_hardirqs_on_caller+0x170/0x18c
ytes in   135.26[  190.182461]  ? trace_hardirqs_on_thunk+0x1a/0x1c
72 ms[  190.182461]  SyS_accept+0x10/0x12
[  190.182461]  entry_SYSCALL_64_fastpath+0x1f/0xc2
[  190.182461] RIP: 0033:0x556b654871e8

[  190.182461] RSP: 002b:00007fed035e3d38 EFLAGS: 00000246 ORIG_RAX: 000000000000002b
[  190.182461] RAX: ffffffffffffffda RBX: ffffffffb7128a40 RCX: 0000556b654871e8
[  190.182461] RDX: 000000c420086b08 RSI: 000000c420086b10 RDI: 0000000000000003
[  190.865562] RBP: ffff949b064b7f98 R08: 0000000000000000 R09: 0000000000000000
2017/05/25 09:58[  190.865562] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000046
:25 [01891] ECHO[  190.865562] R13: 0000000000000006 R14: 0000000000000000 R15: 00000000000000d3
ED:      45875 b[  190.865562]  ? trace_hardirqs_off_caller+0x37/0xa3
ytes in   142.83[  191.214531] ---[ end trace 3271b28a1351ac22 ]---
68 ms
2017/05/25 09:58:26 [01892] ECHOED:       5329 bytes in     0.1535 ms
2017/05/25 09:58:26 [01893] Copy(): read hvsock:4: file already closed
2017/05/25 09:58:26 [01893] Close(): invalid argument
[  191.378324]
[  191.387657] =========================
[  191.387657] [ BUG: held lock freed! ]
2017/05/25 09:58[  191.387657] 4.11.2-linuxkit #1 Tainted: G        W
:26 [01896] ECHO[  191.387657] -------------------------
ED:          0 b[  191.387657] virtsock_stress/655 is freeing memory ffff88627cae6040-ffff88627cae6727, with a lock still held there!
ytes in     0.01[  191.387657]  (sk_lock-AF_VSOCK){+.+.+.}, at: [<ffffffffb7a65704>] lock_sock+0xb/0xd
89 ms[  191.387657] 1 lock held by virtsock_stress/655:
[  191.776424]  #0:  (sk_lock-AF_VSOCK){+.+.+.}, at: [<ffffffffb7a65704>] lock_sock+0xb/0xd
[  191.776424]
[  191.776424] stack backtrace:

[  191.776424] CPU: 1 PID: 655 Comm: virtsock_stress Tainted: G        W       4.11.2-linuxkit #1
[  191.776424] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v2.0 08/26/2016
[  191.776424] Call Trace:
[  192.073482]  dump_stack+0x82/0xb8
[  192.073482]  debug_check_no_locks_freed+0xfe/0x139
[  192.073482]  kmem_cache_free+0x40/0x180
[  192.073482]  __sk_destruct+0x16d/0x186
[  192.073482]  sk_destruct+0x35/0x38
[  192.073482]  __sk_free+0x82/0x9c
[  192.073482]  sk_free+0x17/0x19
[  192.073482]  sock_put+0x12/0x14
[  192.073482]  vsock_accept+0x188/0x299
[  192.073482]  ? add_wait_queue+0x44/0x44
[  192.073482]  SYSC_accept4+0x127/0x1ea
[  192.073482]  ? entry_SYSCALL_64_fastpath+0x5/0xc2
[  192.073482]  ? trace_hardirqs_on_caller+0x170/0x18c
[  192.073482]  ? trace_hardirqs_on_thunk+0x1a/0x1c
[  192.073482]  SyS_accept+0x10/0x12
[  192.073482]  entry_SYSCALL_64_fastpath+0x1f/0xc2
[  192.073482] RIP: 0033:0x556b654871e8
[  192.073482] RSP: 002b:00007fed035e3d38 EFLAGS: 00000246 ORIG_RAX: 000000000000002b
[  192.073482] RAX: ffffffffffffffda RBX: ffffffffb7128a40 RCX: 0000556b654871e8
[  192.073482] RDX: 000000c4200873d8 RSI: 000000c4200873e0 RDI: 0000000000000003
[  192.073482] RBP: ffff949b064b7f98 R08: 0000000000000000 R09: 0000000000000000
[  192.073482] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000046
[  192.073482] R13: 00000000ffffffbb R14: 0000000000000200 R15: 0000000000000100
[  192.073482]  ? trace_hardirqs_off_caller+0x37/0xa3
2017/05/25 09:58:27 [01898] ECHOED:          0 bytes in     0.0126 ms
2017/05/25 09:58:27 [01899] ECHOED:          0 bytes in     0.0099 ms
2017/05/25 09:58:27 [01897] ECHOED:          0 bytes in     0.0083 ms
2017/05/25 09:58:27 [01895] ECHOED:          0 bytes in     0.0083 ms
2017/05/25 09:58:27 [01894] ECHOED:          0 bytes in     0.0292 ms
[...]
2017/05/25 09:58:33 [02546] ECHOED:          0 bytes in     0.1821 ms
2017/05/25 09:58:33 [02547] ECHOED:          0 bytes in     2.7758 ms
2017/05/25 09:58:33 [02548] ECHOED:          0 bytes in     8.0605 ms
[  214.052634] BUG: unable to handle kernel NULL pointer dereference at 0000000000000028
[  214.066101] IP: hvs_release+0x14/0x2b
[  214.066101] PGD 0
[  214.066101]
[  214.066101] Oops: 0000 [#1] SMP
[  214.066101] Modules linked in:
[  214.066101] CPU: 1 PID: 1202 Comm: virtsock_stress Tainted: G        W       4.11.2-linuxkit #1
[  214.066101] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v2.0 08/26/2016
[  214.066101] task: ffff88626659a240 task.stack: ffff949b084fc000
[  214.066101] RIP: 0010:hvs_release+0x14/0x2b
[  214.066101] RSP: 0018:ffff949b084ffdd8 EFLAGS: 00010286
[  214.066101] RAX: 0000000000000000 RBX: ffff8862680b5100 RCX: 0000000000000000
[  214.066101] RDX: ffff8862662515d0 RSI: ffff886266251400 RDI: ffff8862680b5100
[  214.066101] RBP: ffff949b084ffde0 R08: 0000000000000000 R09: 0000000000000000
[  214.066101] R10: ffff949b084ffe78 R11: ffff886266251410 R12: ffff8862680b54d0
[  214.066101] R13: ffff886267d7b4f0 R14: ffff88627e6f08e0 R15: ffff886267d7b4f0
[  214.066101] FS:  00007fed03597eb0(0000) GS:ffff886269a00000(0000) knlGS:0000000000000000
[  214.066101] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  214.066101] CR2: 0000000000000028 CR3: 000000007f49a000 CR4: 00000000003406a0
[  214.066101] Call Trace:
[  214.066101]  __vsock_release+0x2d/0xf9
[  214.066101]  vsock_release+0x16/0x29
[  214.066101]  sock_release+0x1f/0x78
[  214.066101]  sock_close+0x12/0x16
[  214.066101]  __fput+0x10b/0x1af
[  214.066101]  ____fput+0xe/0x10
[  214.066101]  task_work_run+0x77/0xa0
[  214.066101]  prepare_exit_to_usermode+0xaf/0xcc
[  214.066101]  syscall_return_slowpath+0x16f/0x1d3
[  214.066101]  entry_SYSCALL_64_fastpath+0xc0/0xc2
[  214.066101] RIP: 0033:0x556b654245a4
[  214.066101] RSP: 002b:000000c4200c0cd8 EFLAGS: 00000246 ORIG_RAX: 0000000000000003
[  214.066101] RAX: 0000000000000000 RBX: 0000000000000000 RCX: 0000556b654245a4
[  214.066101] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000004
[  214.066101] RBP: 000000c4200c0d20 R08: 0000000000000000 R09: 0000000000000000
[  214.066101] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000000
[  214.066101] R13: 0000000000000006 R14: 0000000000000000 R15: 00000000000000fd
[  214.066101] Code: 00 40 80 e6 02 74 0d 55 48 89 e5 e8 88 ff ff ff 31 c0 5d c3 31 c0 c3 0f 1f 44 00 00 55 48 89 e5 53 48 8b 87 e0 06 00 00 48 89 fb <48> 83 78 28 00 74 05 e8 61 ff ff ff 48 89 df e8 4c cf ff ff 5b
[  214.066101] RIP: hvs_release+0x14/0x2b RSP: ffff949b084ffdd8
[  214.066101] CR2: 0000000000000028
[  214.066101] ---[ end trace 3271b28a1351ac23 ]---
[  214.066101] BUG: sleeping function called from invalid context at ./include/linux/percpu-rwsem.h:33
[  214.066101] in_atomic(): 0, irqs_disabled(): 1, pid: 1202, name: virtsock_stress
[  214.066101] INFO: lockdep is turned off.
[  216.266694] irq event stamp: 332506
[  216.297654] hardirqs last  enabled at (332505): [<ffffffffb70528a8>] get_user_pages_fast+0xf6/0x16c
[  216.297654] hardirqs last disabled at (332506): [<ffffffffb7a78264>] __schedule+0x154/0x6d9
[  216.297654] softirqs last  enabled at (332484): [<ffffffffb7a8094b>] __do_softirq+0x32b/0x434
[  216.545380] softirqs last disabled at (332421): [<ffffffffb70edb6f>] irq_exit+0x5b/0xab
[  216.545380] CPU: 1 PID: 1202 Comm: virtsock_stress Tainted: G      D W       4.11.2-linuxkit #1
[  216.545380] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v2.0 08/26/2016
[  216.545380] Call Trace:
[  216.816084]  dump_stack+0x82/0xb8
[  216.816084]  ___might_sleep+0x1b7/0x1ca
[  216.816084]  __might_sleep+0x77/0x7e
[  216.816084]  exit_signals+0x35/0x18f
[  216.816084]  ? task_work_run+0x77/0xa0
[  216.988188]  do_exit+0x12d/0xa3a
[  217.004083]  rewind_stack_do_exit+0x17/0x20
[  217.048860] RIP: 0033:0x556b654245a4
[  217.076234] RSP: 002b:000000c4200c0cd8 EFLAGS: 00000246 ORIG_RAX: 0000000000000003
[  217.141538] RAX: 0000000000000000 RBX: 0000000000000000 RCX: 0000556b654245a4
[  217.197351] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000004
[  217.293923] RBP: 000000c4200c0d20 R08: 0000000000000000 R09: 0000000000000000
[  217.383797] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000000
[  217.467989] R13: 0000000000000006 R14: 0000000000000000 R15: 00000000000000fd
[  217.550080] BUG: unable to handle kernel NULL pointer dereference at 0000000000000098
[  217.616138] IP: release_sock+0x34/0xa5
[  217.616138] PGD 0
[  217.616138]
[  217.616138] Oops: 0000 [#2] SMP
[  217.616138] Modules linked in:
[  217.616138] CPU: 0 PID: 534 Comm: virtsock_stress Tainted: G      D W       4.11.2-linuxkit #1
[  217.616138] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V UEFI Release v2.0 08/26/2016
[  217.616138] task: ffff886263b7a180 task.stack: ffff949b02ae0000
[  217.616138] RIP: 0010:release_sock+0x34/0xa5
[  217.616138] RSP: 0018:ffff949b02ae3d08 EFLAGS: 00010246
[  217.616138] RAX: 0000000000000000 RBX: ffff8862680b5100 RCX: 0000000000000000
[  217.616138] RDX: ffffffffb78148ba RSI: ffffffffb78148ba RDI: 0000000000000246
[  217.616138] RBP: ffff949b02ae3d18 R08: 0000000000000001 R09: 0000000000000000
[  217.616138] R10: ffff949b02ae3c70 R11: ffff886263b7a180 R12: ffff8862680b5188
[  217.616138] R13: 0000000000000004 R14: ffff949b02ae3dc8 R15: 0000000000000004
[  217.616138] FS:  00007fed037f3eb0(0000) GS:ffff886269800000(0000) knlGS:0000000000000000
[  217.616138] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  217.616138] CR2: 0000000000000098 CR3: 000000007f49a000 CR4: 00000000003406b0
[  217.616138] Call Trace:
[  217.616138]  vsock_stream_sendmsg+0x2c7/0x2ed
[  217.616138]  ? expand_files+0x26c/0x26c
[  217.616138]  ? add_wait_queue+0x44/0x44
[  217.616138]  sock_sendmsg+0x33/0x44
[  217.616138]  sock_write_iter+0x87/0xb3
[  217.616138]  __vfs_write+0x8b/0xc8
[  217.616138]  vfs_write+0xda/0x100
[  217.616138]  SyS_write+0x5f/0xa3
[  217.616138]  entry_SYSCALL_64_fastpath+0x1f/0xc2
[  217.616138] RIP: 0033:0x556b654245a4
[  217.616138] RSP: 002b:000000c420557bb0 EFLAGS: 00000246 ORIG_RAX: 0000000000000001
[  217.616138] RAX: ffffffffffffffda RBX: 0000000000000030 RCX: 0000556b654245a4
[  217.616138] RDX: 0000000000000004 RSI: 000000c420557dbc RDI: 00000000000000a4
[  217.616138] RBP: 00007fed037f3d80 R08: 0000000000000000 R09: 0000000000000000
[  217.616138] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000000
[  217.616138] R13: 0000000000000004 R14: 0000000000000000 R15: 00000000000000da
[  217.616138] Code: 41 54 4c 8d a7 88 00 00 00 53 48 89 fb 4c 89 e7 e8 aa 77 26 00 48 83 bb 38 02 00 00 00 74 08 48 89 df e8 fa fe ff ff 48 8b 43 28 <48> 8b 80 98 00 00 00 48 85 c0 74 05 48 89 df ff d0 83 bb d0 00
[  217.616138] RIP: release_sock+0x34/0xa5 RSP: ffff949b02ae3d08
[  217.616138] CR2: 0000000000000098
[  217.616138] ---[ end trace 3271b28a1351ac24 ]---
[  217.616138] Kernel panic - not syncing: Fatal exception in interrupt
[  217.616138] Kernel Offset: 0x36000000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[  217.616138] ---[ end Kernel panic - not syncing: Fatal exception in interrupt
```

**Comment by dcui** [comment]
@rneugeba 
Hi Rolf,
I'm sorry for the delay -- it turns out I had to deal with a bunch of tricky race conditions, but finally I think I should have resolved all of them; at least I can't crash Linux VM any more with your hvstress test program (I tried 3 million connections with 10 concurrent threads).

The patchset is on this branch:
https://github.com/dcui/linux/commits/decui/hv_sock/v4.11/20170511-debug-0605

I need to refine some of the patches a little, but the changes should be cosmetic.

Can you please try the patchset too?  I'd really love to know how it works for you. 
(The patchset is for v4.11. I'll be rebasing it to the latest linux-next.)

**Comment by rneugeba** [comment]
@dcui thanks for the update.

I finally had some time to test the new patches and they seem to fix the issue. Have been running the stress test on two machines with different Windows builds (`16199.1000` and `14393.1198`) and a 4.11.4 kernel with additional debug enabled. Both passed more then 50k connections. So far so good but I will continue running the tests with some other variations of parameters.

I can't/don't want to integrate the current patches into our tree because https://github.com/dcui/linux/commit/29e6c6204845176c78c7840377a72389d188563c is missing a `Signed-off-by` message.

If you plan to submit these, I strongly suspect you also need a detailed commit message on the above commit as well as on https://github.com/dcui/linux/commit/24ff598cdf98290b25805219eff78336bc08e5ab and https://github.com/dcui/linux/commit/9a00fd7c4ad9c5d1da39c3c44328145fe063ceed

**Comment by dcui** [comment]
Thanks for testing the patchset!  Good to know it works in your side.  

Surely I need to fix the patches by adding Signed-off-by, comments, changelogs, etc.

Actually I may need to re-work the patchset, because there is still a tricky race condition to be resolved (I found it yesterday): if the host closes the connection too quickly, a race in the patches can cause infinite sleeping in vmbus_open().  I'm trying to figure out how to fix this.

**Comment by rn** [comment]
@dcui Actually, on the run I did over night both hang after a while and after hitting `Ctrl-C` on the Linux side I got the error below (this is with a debug kernel):
```
[81331.820761] =============================================
[81331.820761] [ INFO: possible recursive locking detected ]
[81331.820761] 4.11.4-linuxkit #1 Not tainted
[81331.820761] ---------------------------------------------
[81331.820761] virtsock_stress/558 is trying to acquire lock:
[81331.820761]  (sk_lock-AF_VSOCK){+.+.+.}, at: [<ffffffff8ea85039>] lock_sock+0xb/0xd
[81331.820761]
[81331.820761] but task is already holding lock:
[81331.820761]  (sk_lock-AF_VSOCK){+.+.+.}, at: [<ffffffff8ea85039>] lock_sock+0xb/0xd
[81331.820761]
[81331.820761] other info that might help us debug this:
[81331.820761]  Possible unsafe locking scenario:
[81331.820761]
[81331.820761]        CPU0
[81331.820761]        ----
[81331.820761]   lock(sk_lock-AF_VSOCK);
[81331.820761]   lock(sk_lock-AF_VSOCK);
[81331.820761]
[81331.820761]  *** DEADLOCK ***
[81331.820761]
[81331.820761]  May be due to missing lock nesting notation
[81331.820761]
[81331.820761] 1 lock held by virtsock_stress/558:
[81331.820761]  #0:  (sk_lock-AF_VSOCK){+.+.+.}, at: [<ffffffff8ea85039>] lock_sock+0xb/0xd
[81331.820761]
[81331.820761] stack backtrace:
[81331.820761] CPU: 1 PID: 558 Comm: virtsock_stress Not tainted 4.11.4-linuxkit #1
[81331.820761] Hardware name: Microsoft Corporation Virtual Machine/Virtual Machine, BIOS Hyper-V                                               UEFI Release v2.0 08/26/2016
[81331.820761] Call Trace:
[81331.820761]  dump_stack+0x82/0xb8
[81331.820761]  __lock_acquire+0x89c/0xf48
[81331.820761]  ? native_sched_clock+0x1a/0x35
[81331.820761]  ? native_sched_clock+0x1a/0x35
[81331.820761]  ? paravirt_sched_clock+0x9/0xd
[81331.820761]  ? sched_clock+0x9/0xb
[81331.820761]  ? sched_clock_local+0x11/0x7f
[81331.820761]  lock_acquire+0x128/0x1c1
[81331.820761]  ? lock_acquire+0x128/0x1c1
[81331.820761]  ? lock_sock+0xb/0xd
[81331.820761]  lock_sock_nested+0x7d/0x9a
[81331.820761]  ? lock_sock+0xb/0xd
[81331.820761]  lock_sock+0xb/0xd
[81331.820761]  __vsock_release+0x31/0xbf
[81331.820761]  __vsock_release+0xa0/0xbf
[81331.820761]  vsock_release+0x16/0x29
[81331.820761]  sock_release+0x1f/0x78
[81331.820761]  sock_close+0x12/0x16
[81331.820761]  __fput+0x10b/0x1af
[81331.820761]  ____fput+0xe/0x10
[81331.820761]  task_work_run+0x77/0xa0
[81331.820761]  do_exit+0x516/0xa3a
[81331.820761]  ? _raw_spin_unlock_irq+0x2c/0x3b
[81331.820761]  do_group_exit+0x4a/0xb5
[81331.820761]  get_signal+0x563/0x601
[81331.820761]  do_signal+0x37/0x563
[81331.820761]  ? do_futex+0xd2/0x89b
[81331.820761]  ? trace_hardirqs_on+0xd/0xf
[81331.820761]  ? _raw_spin_unlock_irq+0x31/0x3b
[81331.820761]  ? __schedule+0x5ad/0x6d9
[81331.820761]  ? prepare_exit_to_usermode+0x78/0xcc
[81331.820761]  prepare_exit_to_usermode+0x94/0xcc
[81331.820761]  syscall_return_slowpath+0x16f/0x1d3
[81331.820761]  entry_SYSCALL_64_fastpath+0xc0/0xc2
[81331.820761] RIP: 0033:0x55f105453d33
[81331.820761] RSP: 002b:00007fef3b314c10 EFLAGS: 00000286 ORIG_RAX: 00000000000000ca
[81331.820761] RAX: fffffffffffffe00 RBX: 0000000000000000 RCX: 000055f105453d33
[81331.820761] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 000000c4202ea910
[81331.820761] RBP: 00007fef3b314c58 R08: 0000000000000000 R09: 0000000000000000
[81331.820761] R10: 0000000000000000 R11: 0000000000000286 R12: 0000000000000000
[81331.820761] R13: 000000c4202ecd00 R14: 000055f10543ba50 R15: 000055f1057573c0
```

**Comment by dcui** [comment]
Thanks for reporting this calltrace! Can you please share the parameters passed to the programs? Let me try to reproduce this call-trace. I didn't see it before.

**Comment by rn** [comment]
I used the command line reported in the original post, ie -p 10 and 1 million connections

**Comment by dcui** [comment]
Thanks. Let me dig into the issue.

**Comment by rn** [comment]
with the latest patches from https://github.com/dcui/linux/commits/decui/hv_sock/v4.11/20170511-debug-0628-with-signed-off-by-of-dexuan-fixed I have run the stress tests with 10 concurrent connections with 100k connections on two machines (Windows 10 Pro builds 16237.1001 and 14393.1358) without kernel crashes. I started even longer runs now, but once #2192 is merged, this issue will be resolved.

@dcui thanks for the updated patches

**Comment by dcui** [comment]
Thank you Rolf for the help all the way!
Please feel free to contact me if you find any new issue. I'm ready to debug it.
Currently I'm working on how to upstream the patches. I'll keep you updated once I have any progress.

**Comment by dcui** [comment]
(I just typed the comment in another issue in https://github.com/linuxkit/linuxkit/pull/2192 by mistake, but I have deleted the comment in that link.)

**Comment by rn** [comment]
@dcui Happy to:
- test reworked patches and added `Tested-by` to submissions
- review patches before submission

**Comment by dcui** [comment]
@rn Thanks! Will keep you updated.
